minetest/minetest
# Fix pathfinder bugs: returning nil frequently, broken A*, jump through solid nodes

## Goal
Fix the biggest bugs of the pathfinder (`minetest.find_path`).

## Changes
* Fix failure if start or end pos are hovering a few nodes above air (#1701):
    * NOTE: I consider this to be an **important bugfix**. #1701 causes `minetest.find_path` to fail very frequently in common scenarios
    * If startpos is over air, pathfinder will first simulate a drop of `maxdrop` to surface. If surface was found, pos above that will become the start of the algorithm 
    * If destinationpos is over air, pathfinder will also simulate a drop, but with using `maxjump` as limit. If surface was found, pos above that will become the end of the algorithm
    * If startpos and/or destinationpos have been manipulated by the method above, the original startpos/destinationpos will be preserved in the end result, so there will be no ugly surprises on the Lua side
* Fix jumping going through solid matter
* Complete new implementation of `A*` / `A*_noprefetch`. It is now the *real* A* algorithm, not the fake one. The hilariously inefficient paths of Fake A* are gone
* Update documentation of `minetest.find_path`

## Notes
The so-called A* algorithm was completely bonkers (#9331) and also it wasn't even real A*. The offending function was `updateCostHeuristic`. This function is supposedly A* but it's actually not A* at all, it's a greedy recursive non-optimal algorithm, there's not even a concept of an open or closed list. If I understood the code correctly, neighbor nodes are usually just expanded immediately instead of putting them into a list. That's totally not A*. Often generating sometimes very long-winded paths instead of the optimal path. So yeah, completely bonkers.
Thus I decided to replace the function completely.

The new implementation of A* as well as `buildPath` is non-recursive (is there even a recursive A*?). The only function that is still recursive is `updateCostAll` (=Dijkstra).

Question: What is better?
Using the current heuristic function, that is, Manhattan distance on XZ plane only?
Or is it Manhattan distance on all XYZ coords?
In my tests, both function properly, but XYZ has tendency to go up and down on difficult terrain.
Currently, I did not touch the heuristic function.

## Resolved issues (hopefully)
* #1701
* #4198
* #9331 (@SmallJoker: this one is currently NOT on the “list of pathfinder issues” issue #8642)

## To do

It seems to work nicely so far, but needs more testing.

The code is not well-optimized and cleaned up yet, I appreciate any review.
I plan to switch the data structures of the A* search to boost performance, and maybe also seek out redundancies …
Currently, the way how `updateCostHeuristic` keeps track of open and closed list is kind of stupid by redundantly saving a boolean, but it works. I haven't figured out how to do it properly. `openList` is a set, but I think it should be a priority queue.

Known old bug (not caused by PR): Start/end position is one lower than it should be if Y < 0. Probably bad rounding error?

Please do NOT request any features, or this PR will blow up. Direct such comments to one of the pathfinder issues instead, or open a new issue. Concentrate on bugs in this PR, oddities, bad code, bad efficiency, etc.

After this PR, there's obviously still room for new fancy features. But before we even think about them, pathfinder must first become stable, that's very important.

## How to test

Download the `devtest` “game” and use the Pathfinder Tester (`testpathfinder:testpathfinder`) to play around with the pathfinder. Try everything to break it as hard as you can. ;)
Try to reproduce the listed issues. Test `A*` and `A*_noprefetch` extensively. Compare behaviour of previous version.

Try to see what happens if …
* goal or target or both are buried
* goal or target or both are above air (but within `maxdrop`/`maxjump` distance)
* you build a structure that provokes jumping, but something's in the way
* you are creative with your tests ;-)

Usage of Pathfinder Tester: Rightclick places target, punching searches path, sneak-punching changes. Note that jump height, drop height, search distance are hardcoded, so you need to get your hands dirty to change them.
